pkg/ocm/servicelog.go

```diff
@@ -0,0 +1,88 @@
+package ocm
+
+import (
+	"fmt"
+	"regexp"
+
+	"github.com/openshift/ocm-agent-operator/api/v1alpha1"
+	"github.com/openshift/ocm-agent/pkg/consts"
+	"github.com/prometheus/alertmanager/template"
+)
+
+const (
+	ServiceLogActivePrefix  = "Issue Notification"
+	ServiceLogResolvePrefix = "Issue Resolution"
+)
+
+type ServiceLog struct {
+	ServiceName   string                               `json:"service_name"`
+	ClusterUUID   string                               `json:"cluster_uuid,omitempty"`
+	Summary       string                               `json:"summary"`
+	Description   string                               `json:"description"`
+	InternalOnly  bool                                 `json:"internal_only"`
+	Severity      v1alpha1.NotificationSeverity        `json:"severity"`
+	LogType       string                               `json:"log_type,omitempty"`
+	DocReferences []v1alpha1.NotificationReferenceType `json:"doc_references,omitempty"`
+}
+
+func NewServiceLog(summary, firingDesc, resolveDesc, clusterID string, severity v1alpha1.NotificationSeverity, logType string, references []v1alpha1.NotificationReferenceType, firing bool) *ServiceLog {
+	sl := ServiceLog{
+		ServiceName:   consts.ServiceLogServiceName,
+		ClusterUUID:   clusterID,
+		Summary:       summary,
+		InternalOnly:  false,
+		Severity:      severity,
+		LogType:       logType,
+		DocReferences: references,
+	}
+
+	// Use different Summary and Description for firing and resolved status for an alert
+	if firing {
+		sl.Description = firingDesc
+		sl.Summary = ServiceLogActivePrefix + ": " + summary
+	} else {
+		sl.Description = resolveDesc
+		sl.Summary = ServiceLogResolvePrefix + ": " + summary
+	}
+
+	return &sl
+}
+
+var (
+	slVarRefRe = regexp.MustCompile(`\${[^{}]*}`)
+)
+
+func (sl *ServiceLog) ReplacePlaceHoldersWithAlertLabelsAndAnnotations(alert *template.Alert) error {
+	replacePlaceHoldersInString := func(s string) (string, error) {
+		var err error
+		resolvePlaceHolder := func(placeHolder string) string {
+			if err == nil {
+				key, value, isOk := placeHolder[2:len(placeHolder)-1], "", false
+
+				if value, isOk = alert.Labels[key]; !isOk {
+					if value, isOk = alert.Annotations[key]; !isOk {
+						err = fmt.Errorf("alert has no '%s' label or annotation which could be used to replace place holders in the template", key)
+
+						return placeHolder
+					}
+				}
+				return value
+			}
+
+			return placeHolder
+		}
+
+		return slVarRefRe.ReplaceAllStringFunc(s, resolvePlaceHolder), err
+	}
```

Comment:
**Question:** Is there a good reason to declare this function within `ReplacePlaceHoldersWithAlertLabelsAndAnnotations` and `resolvePlaceHolder` within `replacePlaceHoldersInString`? IMO this is pretty hard to read

Reply:
`resolvePlaceHolder` function changes the `err` variable in `replacePlaceHoldersInString` function.
For that reason `resolvePlaceHolder` function is nested in in `replacePlaceHoldersInString` function.

I would have preferred to avoid this side effect by the signature of `resolvePlaceHolder` has to be `func(string) string` when calling `slVarRefRe.ReplaceAllStringFunc`.

I initially designed the code by using [`FindStringSubmatchIndex`](https://pkg.go.dev/regexp#Regexp.FindStringSubmatchIndex) in a loop but it was not elegantly possible to deal with recursive variable replacement with that method.

I also though of using [`FindAllStringSubmatchIndex`](https://pkg.go.dev/regexp#Regexp.FindAllStringSubmatchIndex), but the code was then needlessly complicated.

Reply:
> resolvePlaceHolder function changes the err variable in replacePlaceHoldersInString function.
For that reason resolvePlaceHolder function is nested in in replacePlaceHoldersInString function.

This part makes sense. But can't we move `replacePlaceHoldersInString` out of `ReplacePlaceHoldersWithAlertLabelsAndAnnotations`?

Reply:
Technically we could, but `replacePlaceHoldersInString` uses the `alert` parameter passed to `ReplacePlaceHoldersWithAlertLabelsAndAnnotations`; if I move it out of it:
-  I will have to add some `alert` parameter to `replacePlaceHoldersInString`
- By doing so I may have to rename the method to tell that the place holders are replaced with the labels & annotations taken from the new `alert` parameter... while this was implicit with the previous nesting. Method name could become `replacePlaceHoldersWithAlertLabelsAndAnnotationsInString`... this is long, not sure this is suitable.